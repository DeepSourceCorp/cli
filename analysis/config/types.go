package config

import "github.com/deepsourcelabs/cli/analysis/lsp"

///////////////////////////////////////////////////////
// DSConfig is the struct for .deepsource.toml file //
/////////////////////////////////////////////////////

type DSConfig struct {
	Version         int           `mapstructure:"version,omitempty" json:"version" toml:"version"`
	ExcludePatterns []string      `mapstructure:"exclude_patterns,omitempty" json:"exclude_patterns,omitempty" toml:"exclude_patterns"`
	TestPatterns    []string      `mapstructure:"test_patterns,omitempty" json:"test_patterns,omitempty" toml:"test_patterns"`
	Analyzers       []Analyzer    `mapstructure:"analyzers,omitempty" json:"analyzers,omitempty" toml:"analyzers"`
	Transformers    []Transformer `mapstructure:"transformers,omitempty" json:"transformers,omitempty" toml:"transformers"`
}

type Analyzer struct {
	Name                string      `mapstructure:"name,omitempty" json:"name,omitempty" toml:"name"`
	RuntimeVersion      string      `mapstructure:"runtime_version,omitempty" json:"runtime_version,omitempty" toml:"runtime_version,omitempty"`
	Enabled             bool        `mapstructure:"enabled,omitempty" json:"enabled" toml:"enabled"`
	DependencyFilePaths []string    `mapstructure:"dependency_file_paths,omitempty" json:"dependency_file_paths,omitempty"`
	Meta                interface{} `mapstructure:"meta,omitempty" json:"meta,omitempty" toml:"meta"`
	Thresholds          interface{} `mapstructure:"thresholds,omitempty" json:"thresholds,omitempty"`
}

type Transformer struct {
	Name    string `mapstructure:"name,omitempty" json:"name,omitempty" toml:"name"`
	Enabled bool   `mapstructure:"enabled,omitempty" json:"enabled,omitempty" toml:"enabled"`
}

///////////////////////////////////////
// LSP based Analysis Config Types  //
/////////////////////////////////////

type AnalysisConfig struct {
	Files           []lsp.TextDocumentItem `json:"files"`
	TestFiles       []lsp.TextDocumentItem `json:"test_files"`
	ExcludedFiles   []lsp.TextDocumentItem `json:"excluded_files"`
	ExcludePatterns []string               `json:"exclude_patterns"`
	TestPatterns    []string               `json:"test_patterns"`
	AnalyzerMeta    interface{}            `json:"analyzer_meta"`
}

//////////////////////////////////////
// LSP based Analysis Result Types  //
/////////////////////////////////////

type Namespace struct {
	Key   string  `json:"key"`
	Value float64 `json:"value"`
}

type Metric struct {
	MetricCode string      `json:"metric_code"`
	Namespaces []Namespace `json:"namespaces"`
}

type AnalysisResult struct {
	Issues   []lsp.Diagnostic `json:"issues"`
	Metrics  []Metric         `json:"metrics,omitempty"`
	IsPassed bool             `json:"is_passed"`
	Errors   []Error          `json:"errors"`
	// Errors    []lsp.Diagnostic `json:"errors"`
	ExtraData interface{} `json:"extra_data"`
}

////////////////////////////////////
// Default Analysis Result Types  //
///////////////////////////////////

type Error struct {
	HMessage string `json:"hmessage"`
	Level    int    `json:"level"`
}

type Coordinate struct {
	Line   int `json:"line"`
	Column int `json:"column"`
}

type Position struct {
	Begin Coordinate `json:"begin"`
	End   Coordinate `json:"end"`
}

type Location struct {
	Path     string   `json:"path"`
	Position Position `json:"position"`
}

type Issue struct {
	Code     string   `json:"issue_code"`
	Title    string   `json:"issue_text"`
	Location Location `json:"location"`
}

type DefaultAnalysisResult struct {
	Issues    []Issue     `json:"issues"`
	Metrics   []Metric    `json:"metrics,omitempty"`
	IsPassed  bool        `json:"is_passed"`
	Errors    []Error     `json:"errors"`
	ExtraData interface{} `json:"extra_data"`
}
