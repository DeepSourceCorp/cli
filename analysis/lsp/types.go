package lsp

//////////////////////////////
// Document identity types  //
/////////////////////////////

type DocumentURI string

type TextDocumentItem struct {
	URI        DocumentURI `json:"uri"`
	LanguageID string      `json:"languageID,omitempty"`
	Version    int         `json:"version,omitempty"`
	Text       string      `json:"text,omitempty"`
}

///////////////////////////////
// Diagnostic related types //
/////////////////////////////

type DiagnosticSeverity int

const (
	Error       DiagnosticSeverity = 1
	Warning     DiagnosticSeverity = 2
	Information DiagnosticSeverity = 3
	Hint        DiagnosticSeverity = 4
)

type Position struct {
	Line      int `json:"line"`
	Character int `json:"character"`
}
type Range struct {
	Start Position `json:"start"`
	End   Position `json:"end"`
}

type Location struct {
	URI   string `json:"uri"`
	Range Range  `json:"range"`
}

type DiagnosticRelatedInformation struct {
	Location Location `json:"location"`
	Message  string   `json:"message"`
}

type Diagnostic struct {
	Range    Range              `json:"range"`
	Severity DiagnosticSeverity `json:"severity,omitempty"`
	Code     string             `json:"code,omitempty"`
	Source   string             `json:"source,omitempty"`
	Message  string             `json:"message"`

	/**
	 * An array of related diagnostic information, e.g. when symbol-names within
	 * a scope collide all definitions can be marked via this property.
	 * var a,b
	 * a := 2
	 * Issues in line 1 and 2 are related.
	 */
	RelatedInformation []DiagnosticRelatedInformation `json:"relatedInformation"`
}
