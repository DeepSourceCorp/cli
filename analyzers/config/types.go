package config

type Analysis struct {
	Command string `toml:"command" json:"command" validate:"required"`
}

// TODO: Remove the comment when we start support for Autofix.
// type Autofix struct {
//     Command string `toml:"command" json:"command" analyzertoml:"required"`
// }

type Build struct {
	Builder    string `toml:"builder" json:"builder" validate:"omitempty"`
	Dockerfile string `toml:"dockerfile" json:"dockerfile,omitempty" validate:"omitempty"`
	Script     string `toml:"script" json:"script" validate:"omitempty"`
}

type URL struct {
	Source        string `toml:"source" json:"source" validate:"omitempty,url"`
	Documentation string `toml:"documentation" json:"documentation,omitempty" validate:"omitempty,url"`
	BugTracker    string `toml:"bug_tracker" json:"bug_tracker" validate:"omitempty,url"`
}

type Test struct {
	Command string `toml:"command" json:"command" validate:"omitempty"`
}

// analyzer.toml type
type AnalyzerMetadata struct {
	// Analyzer specific data
	Name                 string            `toml:"name" json:"name" validate:"required"`
	Shortcode            string            `toml:"shortcode" json:"shortcode" validate:"required"` // New
	Description          string            `toml:"description" json:"description" validate:"required"`
	Category             string            `toml:"category" json:"category" validate:"required"`
	Tags                 []string          `toml:"tags" json:"tags,omitempty" validate:"omitempty"`
	URL                  URL               `toml:"urls" json:"urls,omitempty" validate:"omitempty"`
	EnvironmentVariables map[string]string `toml:"environment_variables" json:"environment_variables,omitempty" validate:"omitempty"`

	// Analyzer, Autofix and Transformer config
	Analysis Analysis `toml:"analysis" json:"analysis" validate:"required"`
	// Autofix Autofix `toml:"autofix" json:"autofix" validate:"omitempty"`

	// Build steps needed to build the analyzer
	Build Build `toml:"build" json:"build" validate:"omitempty"`

	// Test command for running the testing workflow
	Test Test `toml:"test" json:"test" validate:"omitempty"`
}

// Analyzer issue description
type AnalyzerIssue struct {
	Shortcode   string `validate:"omitempty"`
	Title       string `toml:"title" json:"title" validate:"required"`
	Description string `toml:"description" json:"description" validate:"required"`
	Category    string `toml:"category" json:"category" validate:"required"`
}
