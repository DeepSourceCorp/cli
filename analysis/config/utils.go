package config

import (
	"log"
	"os"
	"path"
	"path/filepath"
	"strings"

	"github.com/bmatcuk/doublestar/v4"
)

/* Walks the `CODE_PATH` directory and returns all the files other than the ones present
 * in the .git folder for analysis in the form of a string array */
func readAllFiles(codePath string) ([]string, error) {
	fileCount := 0

	allFiles := make([]string, 0)

	err := filepath.Walk(codePath,
		func(path string, fileInfo os.FileInfo, err error) error {
			if err != nil {
				return err
			}
			fileCount++

			/* Check the following before appending to the list of files:
			 * - Should not be a directory
			 * - The walked file should not be present in .git folder
			 * TODO: Check why fileCount != 1 condition is used here */
			if !fileInfo.IsDir() || !strings.HasPrefix(path, filepath.Join(codePath, ".git")) && fileCount != 1 {
				allFiles = append(allFiles, path)
			}
			return nil
		})
	if err != nil {
		return allFiles, err
	}
	return allFiles, nil
}

// Returns the slice of files matching certain glob patterns
func (r *AnalysisRun) getMatchingFiles(patterns []string) ([]string, error) {
	matchedFiles := make([]string, 0)

	// Return all the files if no exclude_patterns are configured
	if len(patterns) == 0 {
		return matchedFiles, nil
	}

	for _, file := range r.AnalysisFiles {
		for i := range patterns {
			match, err := doublestar.Match(path.Join(r.CodePath, patterns[i]), file)
			if err != nil {
				log.Printf("Failed to match the file %s with pattern %s\n", file, patterns[i])
			}
			if match {
				matchedFiles = append(matchedFiles, file)
				break
			}
		}
	}
	return matchedFiles, nil
}

// Filters the analysis files and removes the files matching the exclude_patterns from them
// TODO: Improve the logic here
func (r *AnalysisRun) filterAnalysisFiles() {
	filteredFiles := []string{}
	for _, file := range r.AnalysisFiles {
		for _, excludedFile := range r.ExcludedFiles {
			if file == excludedFile {
				break
			}
			filteredFiles = append(filteredFiles, file)
		}
	}
	r.AnalysisFiles = filteredFiles
}
