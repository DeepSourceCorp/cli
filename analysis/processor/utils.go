package processor

import (
	"sort"

	"github.com/deepsourcelabs/cli/types"
)

// sortIssuesByFile sorts the issues in an alphabetical order according to the filenames
// where they got reported.
func (p *ProcessAnalysisResults) sortIssuesByFile() {
	sort.Slice(p.AnalysisResult.Issues, func(i, j int) bool {
		return p.AnalysisResult.Issues[i].Location.Path < p.AnalysisResult.Issues[j].Location.Path
	})
}

// Prepare a map with unique filenames as key and the issue range for each file as value
// This is done to ensure fewer loops when processing issues.
type IssueRange struct {
	Filename   string // The file which has the issues
	BeginIndex int    // Array index in report.Issues where the particular issue starts
	EndIndex   int    // Array index in report.Issues where the particular issue ends
}

// GenerateIssueRangeSlice generates an array containing the issue ranges with respect to files
// that helps us to go through them and map them to the files where they got reported instead
// of opening the file for each of them. The generated index looks like this:

// [{analyzer.go 0 0} {autofix_patch.go 1 1} {difftool.go 2 2} {patch.patch 3 5} {proc_skip_cq_test.go 6 6}]

// Here, the first field if filename and the second and third fields are the index range in which the issues reported
// in these files lie in the sorted AnalyzerReport slice.
func createIssueFileRange(report types.AnalysisResult) []IssueRange {
	fileCount := 0 // for 1 file, 0 based indexing
	issuesRange := []IssueRange{}
	prevFilename := report.Issues[0].Location.Path

	issueRange := IssueRange{
		BeginIndex: 0,
		EndIndex:   len(report.Issues) - 1,
		Filename:   prevFilename,
	}
	issuesRange = append(issuesRange, issueRange)

	// Iterating over the issues and creating an array containing issues with index data about
	// the files in which those issues are present
	for i := 1; i < len(report.Issues); i++ {
		issue := report.Issues[i]
		currentFilename := issue.Location.Path

		// TODO: Check when this condition is implied
		if issue.Location.Position.End.Line == -1 {
			issue.Location.Position.End.Line = issue.Location.Position.Begin.Line
		}

		if currentFilename != prevFilename {
			fileCount++

			issueRange = issuesRange[fileCount-1]
			issueRange.EndIndex = i - 1
			issuesRange[fileCount-1] = issueRange

			// Create for the new file
			issueRange = IssueRange{
				Filename:   currentFilename,
				BeginIndex: i,
				EndIndex:   len(report.Issues) - 1,
			}

			issuesRange = append(issuesRange, issueRange)
			prevFilename = currentFilename
		}
	}
	return issuesRange
}

// formatLSPResultsToDefault converts the LSP based analysis results into the default format supported by DeepSource.
func (p *ProcessAnalysisResults) formatLSPResultsToDefault() {
	p.AnalysisResult.IsPassed = p.Report.IsPassed
	p.AnalysisResult.Metrics = append(p.Report.Metrics, p.AnalysisResult.Metrics...)
	p.AnalysisResult.Errors = append(p.Report.Errors, p.AnalysisResult.Errors...)

	// Appending the issues to the default format of Analysis report
	for _, issue := range p.Report.Issues {
		analysisIssue := types.Issue{
			IssueCode: issue.Code,
			IssueText: issue.Message,
			Location: types.Location{
				Path: issue.RelatedInformation[0].Location.URI,
				Position: types.Position{
					Begin: types.Coordinate{
						Line:   issue.Range.Start.Line,
						Column: issue.Range.Start.Character,
					},
					End: types.Coordinate{
						Line:   issue.Range.End.Line,
						Column: issue.Range.End.Character,
					},
				},
			},
		}
		p.AnalysisResult.Issues = append(p.AnalysisResult.Issues, analysisIssue)
	}
}
