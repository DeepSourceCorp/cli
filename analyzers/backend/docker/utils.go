package docker

import (
	"bufio"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"strings"
)

type ErrorLine struct {
	Error       string      `json:"error"`
	ErrorDetail ErrorDetail `json:"errorDetail"`
}

type ErrorDetail struct {
	Message string `json:"message"`
}

type DockerBuildResponse struct {
	Stream string `json:"stream"`
}

func CheckBuildResponse(rd io.Reader, showAllLogs bool) error {
	var lastLine []byte
	count := 0
	var currentStream string

	scanner := bufio.NewScanner(rd)
	for scanner.Scan() {
		lastLine = scanner.Bytes()
		d := &DockerBuildResponse{}
		err := json.Unmarshal(lastLine, d)
		if err != nil {
			return err
		}
		if d.Stream == "" || d.Stream == "\n" || strings.Contains(d.Stream, "--->") || strings.TrimSuffix(d.Stream, "\n") == currentStream {
			continue
		}
		currentStream = strings.TrimSuffix(d.Stream, "\n")
		if showAllLogs {
			fmt.Println(currentStream)
		}
		count++
	}

	errLine := &ErrorLine{}
	json.Unmarshal([]byte(lastLine), errLine)
	if errLine.Error != "" {
		return errors.New(errLine.Error)
	}
	return scanner.Err()
}


