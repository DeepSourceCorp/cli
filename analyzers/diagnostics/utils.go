package diagnostics

import (
	"fmt"
	"os"
	"regexp"
	"strings"

	"github.com/morikuni/aec"
)

// fieldRegexp is a regular expression for checking the current line is a field or not.
var fieldRegexp = `\s*(?P<field>.+)\s*=\s*`

// readFileContent reads the file and returns its content.
func readFileContent(filename string) (string, error) {
	content, err := os.ReadFile(filename)
	if err != nil {
		return "", err
	}
	return string(content), nil
}

// prepareCodeFrame prepares a code frame using the file content. The code frame is meant to be displayed on the console while reporting diagnostics.
// NOTE: lineNum always starts from 0.
func prepareCodeFrame(lineNum int, lines []string) string {
	frame := ""

	if lineNum <= 1 {
		// Case 1: When the line is near the top of the file.
		// Generate a frame with the current and next line only.
		frame += aec.LightRedF.Apply(fmt.Sprintf("> %d | %s\n", lineNum+1, lines[lineNum]))
		frame += fmt.Sprintf("  %d | %s\n", lineNum+2, lines[lineNum+1])

	} else if lineNum >= (len(lines) - 1) {
		// Case 2: When the line is near the bottom of the file.
		// Generate a frame with the current line only.
		frame += aec.LightRedF.Apply(fmt.Sprintf("> %d | %s\n", lineNum, lines[lineNum-1]))
	} else {
		// Case 3: When the line is in between the top and the bottom.
		// Generate a frame with the the previous, current and the next line.
		frame += fmt.Sprintf("  %d | %s\n", lineNum, lines[lineNum-1])
		frame += aec.LightRedF.Apply(fmt.Sprintf("> %d | %s\n", lineNum+1, lines[lineNum]))
		frame += fmt.Sprintf("  %d | %s\n", lineNum+2, lines[lineNum+1])
	}

	return frame
}

// checkField checks if the line contains the field.
func checkField(line, field string) bool {
	// We use fieldRegexp for checking if the current line is a field or not.
	// If it's not a field, for example, comments for explanation, etc., then checkField should return false.
	exp := regexp.MustCompile(fieldRegexp)

	// Get index for the "field" named group.
	matches := exp.FindStringSubmatch(line)
	groupIndex := exp.SubexpIndex("field")

	// Capture match. We proceed only if the groupIndex isn't -1 (group not found), and if the number of matches aren't zero.
	// Once done, we populate match with the match found using the regular expression.
	match := ""
	if len(matches) != 0 && groupIndex != -1 {
		match = matches[groupIndex]
	}
	match = strings.TrimSpace(match)

	// Return true if the match contains the field.
	return strings.Contains(match, field)
}
