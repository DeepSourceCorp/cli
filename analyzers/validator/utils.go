package validator

import (
	"errors"
	"fmt"
	"path"
	"reflect"
	"regexp"
	"strings"

	validate "github.com/go-playground/validator/v10"
	"github.com/pelletier/go-toml/v2"
)

// codeframeFieldRegexp is a regular expression used for matching a TOML field present in the codeframe generated by go-toml.
var codeframeFieldRegexp = `\d\s*[|].(?P<field>\w+)\s*=\s*\S*`

// Returns the list of required fields from the error message returned by the `go-playground/validator` library
func getMissingRequiredFields(err error, config interface{}) []string {
	missingRequiredFields := []string{}
	errs := err.(validate.ValidationErrors)
	for _, err := range errs {
		if err.Tag() == "required" {
			c := reflect.ValueOf(config)
			for i := 0; i < c.Type().NumField(); i++ {
				if err.Field() == c.Type().Field(i).Name {
					missingRequiredFields = append(missingRequiredFields, c.Type().Field(i).Tag.Get("toml"))
				}
			}
		}
	}
	return missingRequiredFields
}

// Handle decoding errors reported by go-toml
func handleTOMLDecodeErrors(err error, filePath string) *ValidationFailure {
	var usefulResponse, expectedType, receivedType, fieldName, decodeErrorMessage string

	// We check for go-toml's StrictMissingError, as the decoder returns a StrictMissingError when there exists a mismatch between the destination and the extracted fields.
	// go-toml doesn't return the field name associated with the error, however, StrictMissingError returns a human-readable codeframe, which can be used to extract the error-causing fields.
	// Ref: https://pkg.go.dev/github.com/pelletier/go-toml/v2#example-Decoder.DisallowUnknownFields

	var strictMissingError *toml.StrictMissingError
	if errors.As(err, &strictMissingError) {
		// Extract human-readable codeframe.
		codeframe := strictMissingError.String()

		// Retrieve field names from the codeframe.
		fields := parseCodeframe(codeframe)

		// Construct errors using the extracted field names.
		var errors []ErrorMeta
		for _, field := range fields {
			errMeta := ErrorMeta{
				Level:   DecodeErr,
				Field:   field,
				Message: fmt.Sprintf("Invalid field found: %s", field),
			}
			errors = append(errors, errMeta)
		}

		// Construct a validation failure.
		validationError := ValidationFailure{
			File:   filePath,
			Errors: errors,
		}

		return &validationError
	}

	// Get the DecodeError exported by go-toml
	// Ref: https://pkg.go.dev/github.com/pelletier/go-toml/v2#DecodeError
	var decodeErr *toml.DecodeError
	if !errors.As(err, &decodeErr) {
		decodeErrorMessage = err.Error()

		// Handle strict mode error when some alien fields are added in the user configured TOML
		if strings.HasPrefix(err.Error(), "strict mode") {
			decodeErrorMessage = fmt.Sprintf("Failed to parse %s. Invalid fields detected.", path.Base(filePath))
		}
		validationError := ValidationFailure{
			File: filePath,
			Errors: []ErrorMeta{
				{
					Level:   DecodeErr,
					Field:   "",
					Message: decodeErrorMessage,
				},
			},
		}
		return &validationError
	}

	/* =================================================
	 * Extract the data about the decoding failure and return
	 * a validation failure response
	 * ================================================= */

	errorMessage := decodeErr.Error()
	// Error case 1: `toml: cannot decode TOML integer into struct field types.AnalyzerTOML.Name of type string"`
	if strings.HasPrefix(errorMessage, "toml: cannot decode TOML") {

		usefulResponse = strings.TrimPrefix(errorMessage, "toml: cannot decode TOML ")
		responseArray := strings.Split(usefulResponse, " ")

		expectedType = responseArray[len(responseArray)-1]
		receivedType = responseArray[0]
		fieldData := responseArray[len(responseArray)-4]
		index := strings.LastIndex(fieldData, ".")
		fieldName = strings.ToLower(fieldData[index:])
		// Framing the decoding failure error message
		decodeErrorMessage = fmt.Sprintf("expected the field \"%s\" of type %s. Got %s.", fieldName, expectedType, receivedType)

	} else if strings.HasPrefix(errorMessage, "toml: cannot store TOML") {

		// Error case 2: `toml: cannot store TOML string into a Go slice`
		usefulResponse = strings.TrimPrefix(errorMessage, "toml: cannot store TOML ")
		responseArray := strings.Split(usefulResponse, " ")

		expectedType = responseArray[len(responseArray)-1]
		receivedType = responseArray[0]
		decodeErrorMessage = fmt.Sprintf("expected type for one of the fields : %s. Received: %s.", expectedType, receivedType)
	} else {
		decodeErrorMessage = errorMessage
		fieldName = ""
	}

	validationError := ValidationFailure{
		File: filePath,
		Errors: []ErrorMeta{
			{
				Level:   DecodeErr,
				Field:   fieldName,
				Message: decodeErrorMessage,
			},
		},
	}
	return &validationError
}

// parseCodeframe parses the codeframe generated by go-toml, and extracts field names which are associated with the error.
func parseCodeframe(codeframe string) []string {
	var fields []string
	lines := strings.Split(codeframe, "\n")

	// Compile a regular expression for matching fields.
	exp := regexp.MustCompile(codeframeFieldRegexp)

	// This is added as a safety check. pelletier/go-toml/v2 doesn't return codeframes lesser than 2 lines.
	if len(lines) < 2 {
		return []string{""}
	}

	for i := 1; i < len(lines); i++ {
		// Retrieve current and previous line from the codeframe.
		currLine := lines[i]
		prevLine := lines[i-1]

		// Check if the current line contains the "missing field" message.
		if strings.Contains(currLine, "missing field") {
			// If yes, use the codeframeFieldRegexp regular expression for matching the previous line to retrieve the field name.
			matches := exp.FindStringSubmatch(prevLine)

			// Get index for the "field" named group.
			groupIndex := exp.SubexpIndex("field")

			// Capture match. We proceed only if the groupIndex isn't -1 (group not found), and if the number of matches aren't zero.
			match := ""
			if len(matches) != 0 && groupIndex != -1 {
				match = matches[groupIndex]
			}

			// Add the field name to fields.
			field := strings.TrimSpace(match)
			fields = append(fields, field)
		}
	}

	return fields
}
