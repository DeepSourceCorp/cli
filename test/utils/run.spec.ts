import { Pr, PrStatus, Run, RunStatus } from '~/types/types'
import { generalizePR, generalizeRun } from '~/utils/runs'

describe('[[run utils]]', () => {
  const pr: Pr = {
    id: '1234',
    createdAt: 'today',
    modifiedAt: 'never',
    status: PrStatus.Open,
    alive: true,
    baseBranch: 'master',
    branch: 'feat--yesterday',
    number: 1337,
    raisedCount: 420,
    resolvedCount: 101,
    runCount: 78,
    latestAnalysisRun: {
      id: 'lolrun',
      runId: '12356',
      commitOid: 'o0o0o0o0o',
      status: RunStatus.Pass,
      createdAt: 'day before yesterday',
      modifiedAt: 'day after tomorrow',
      checks: { pageInfo: { hasNextPage: false, hasPreviousPage: true }, edges: [] },
      extraData: {}
    }
  }

  const run: Run = {
    id: '1234',
    branchName: 'feat--yesterday',
    runId: '12356',
    branchRunCount: 78,
    pullRequestNumberDisplay: '#1337',
    commitOid: 'o0o0o0o0o',
    createdAt: 'day before yesterday',
    modifiedAt: 'day after tomorrow',
    issuesRaisedCount: 420,
    issuesResolvedCount: 101,
    status: RunStatus.Pass,
    commitMessage: 'feat: add yesterday',
    checks: { pageInfo: { hasNextPage: false, hasPreviousPage: true }, edges: [] },
    extraData: {}
  }

  test('generalizeRun', () => {
    const generalizedRun = generalizeRun(run)
    expect(generalizedRun).toMatchSnapshot('generalizeRun')
  })

  test('generalizePR', () => {
    const generalizedRun = generalizePR(pr)
    expect(generalizedRun).toMatchSnapshot('generalizePR')
  })
})
