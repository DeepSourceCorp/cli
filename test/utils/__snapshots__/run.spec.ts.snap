// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[[run utils]] generalizePR: generalizePR 1`] = `
Object {
  "branchName": "feat--yesterday",
  "commitOid": "o0o0o0o0o",
  "createdAt": "day before yesterday",
  "issuesRaisedCount": 420,
  "issuesResolvedCount": 101,
  "prNumber": "#1337",
  "runCount": 78,
  "runId": "12356",
  "status": "PASS",
  "title": undefined,
}
`;

exports[`[[run utils]] generalizeRun: generalizeRun 1`] = `
Object {
  "branchName": "feat--yesterday",
  "commitOid": "o0o0o0o0o",
  "createdAt": "day before yesterday",
  "issuesRaisedCount": 420,
  "issuesResolvedCount": 101,
  "prNumber": "#1337",
  "runCount": 78,
  "runId": "12356",
  "status": "PASS",
  "title": "feat: add yesterday",
}
`;
